include/SampleFrame.h

```diff
@@ -0,0 +1,234 @@
+
+#include <cmath>
+#include <algorithm>
+
+
+namespace lmms
+{
+
+class SampleFrame
+{
+public:
+	SampleFrame() : SampleFrame(0., 0.)
+	{
+	}
+
+	SampleFrame(sample_t value) : SampleFrame(value, value)
+	{
+	}
+
+	SampleFrame(sample_t left, sample_t right) :
+		m_samples({ left, right })
+	{
+	}
+
+	sample_t* data()
+	{
+		return m_samples.data();
+	}
+
+	const sample_t* data() const
+	{
+		return m_samples.data();
+	}
+
+	sample_t& left()
+	{
+		return m_samples[0];
+	}
+
+	const sample_t& left() const
+	{
+		return m_samples[0];
+	}
+
+	void setLeft(const sample_t& value)
+	{
+		m_samples[0] = value;
+	}
+
+	sample_t& right()
+	{
+		return m_samples[1];
+	}
+
+	const sample_t& right() const
+	{
+		return m_samples[1];
+	}
+
+	void setRight(const sample_t& value)
+	{
+		m_samples[1] = value;
+	}
+
+	sample_t& operator[](size_t index)
+	{
+		return m_samples[index];
+	}
+
+	const sample_t& operator[](size_t index) const
+	{
+		return m_samples[index];
+	}
+
+	SampleFrame& operator=(float v)
+	{
+		left() = v;
+		right() = v;
+
+		return *this;
+	}
+
+	SampleFrame operator+(const SampleFrame& other) const
+	{
+		return SampleFrame(left() + other.left(), right() + other.right());
+	}
+
+	void operator+=(const SampleFrame& other)
+	{
+		auto & l = left();
+		auto & r = right();
+
+		l += other.left();
+		r += other.right();
+	}
+
+	SampleFrame operator*(float value) const
+	{
+		return SampleFrame(left() * value, right() * value);
+	}
+
+	void operator*=(float value)
+	{
+		setLeft(left() * value);
+		setRight(right() * value);
+	}
+
+	SampleFrame operator*(const SampleFrame& other) const
+	{
+		return SampleFrame(left() * other.left(), right() * other.right());
+	}
+
+	void operator*=(const SampleFrame& other)
+	{
+		left() *= other.left();
+		right() *= other.right();
+	}
+
+	sample_t scalarProduct(const SampleFrame& other) const
+	{
+		return left() * other.left() + right() * other.right();
+	}
+
+	SampleFrame abs() const
+	{
+		return SampleFrame{std::abs(this->left()), std::abs(this->right())};
+	}
+
+	void max(const SampleFrame& other)
+	{
+		if (other.left() > left())
+		{
+			setLeft(other.left());
+		}
+
+		if (other.right() > right())
+		{
+			setRight(other.right());
+		}
+	}
+
+	sample_t average() const
+	{
+		return (left() + right()) / 2;
+	}
+
+	void clamp(sample_t low, sample_t high)
+	{
+		auto & l = left();
+		l = std::clamp(l, low, high);
+
+		auto & r = right();
+		r = std::clamp(r, low, high);
+	}
+
+	bool containsInf() const
+	{
+		return std::isinf(left()) || std::isinf(right());
+	}
+
+	bool containsNaN() const
+	{
+		return std::isnan(left()) || std::isnan(right());
+	}
+
+private:
+	std::array<sample_t, DEFAULT_CHANNELS> m_samples;
+};
+
+inline void zeroSampleFrames(SampleFrame* buffer, fpp_t frames)
+{
+	// The equivalent of the following operation which yields compiler warnings
+	// memset(buffer, 0, sizeof(SampleFrame) * frames);
+
+	for (fpp_t i = 0; i < frames; ++i)
+	{
+		buffer[i] = SampleFrame();
+	}
```

Comment:
Could use `std::fill(buffer, buffer + frames, SampleFrame{});`, which I believe is preferred both to `std::memset` as well as to a manual implementation of the loop.

Reply:
Done with commit 1fe8971e.